### PR TITLE
analyze: implement Box<T> rewrites

### PR DESCRIPTION
**Pull request by spernsteiner**
This branch adds rewrites to convert pointers with the `FREE` permission into `Box<T>`.  This follows the dynamic ownership tracking proposal in #1097.

Specific changes:
* Adds a new `dyn_owned: bool` field to `TypeDesc`, which indicates that the pointer type should be wrapped in a dynamic ownership wrapper.
* Extends `ZeroizeType` handling to support zero-initializing some pointer types (necessary to allow `malloc`/`calloc` of structs that contain pointers).
* Adds support for `Box` and `dyn_owned` related casts in `mir_op` and `rewrite::expr::convert`.
* Adds custom rewrite rules for `malloc`, `calloc`, `free`, and `realloc`.

### CODE REVIEW

File: c2rust-analyze/src/rewrite/expr/convert.rs [hunk 1]
```diff
@@ -286,6 +286,126 @@ impl<'tcx> ConvertVisitor<'tcx> {
...
+                ref zero_ty,
+                elem_size,
+                single,
+            } |
+            mir_op::RewriteKind::CallocSafe {
+                ref zero_ty,
+                elem_size,
+                single,
+            } => {
+                // `malloc(n)` -> `Box::new(z)` or similar
```
Comment (ahomescu) [inline c2rust-analyze/src/rewrite/expr/convert.rs:299]: Could you add an assertion here that the size of `zero_ty` is `elem_size`? Or can `zero_ty` be a different type?

Comment (spernsteiner) [reply c2rust-analyze/src/rewrite/expr/convert.rs:299]: The sizes don't necessarily match.  `elem_size` is the size of the original, unrewritten type, which we use for converting `memcpy`/`memset` byte counts to a number of elements.  `zero_ty` indicates how to zero out the rewritten type, which might have a different size.  (Also, `ZeroizeTy` doesn't actually contain enough detail to allow computing the size.)

Comment (ahomescu) [reply c2rust-analyze/src/rewrite/expr/convert.rs:299]: > `zero_ty` indicates how to zero out the rewritten type, which might have a different size.

Could that cause compiler errors?

File: c2rust-analyze/src/rewrite/expr/convert.rs [hunk 2]
```diff
@@ -286,6 +286,119 @@ impl<'tcx> ConvertVisitor<'tcx> {
...
+                    ],
+                    _ => unreachable!(),
+                };
+                let expr = if single {
+                    stmts.push(Rewrite::Text("assert_eq!(n, 1)".into()));
+                    format_rewrite!("Box::new({})", zeroize_expr)
+                } else {
+                    stmts.push(Rewrite::Let1(
+                        "mut v".into(),
+                        Box::new(Rewrite::Text("Vec::with_capacity(n)".into())),
```
Comment (ahomescu) [inline c2rust-analyze/src/rewrite/expr/convert.rs:326]: Could this just be `vec![{zeroize_expr}; n]`?

File: c2rust-analyze/src/rewrite/expr/convert.rs [hunk 3]
```diff
@@ -566,7 +679,7 @@ fn generate_zeroize_code(zero_ty: &ZeroizeType, lv: &str) -> String {
     match *zero_ty {
         ZeroizeType::Int => format!("{lv} = 0"),
         ZeroizeType::Bool => format!("{lv} = false"),
-        ZeroizeType::Iterable(ref elem_zero_ty) => format!(
```
Comment (ahomescu) [inline c2rust-analyze/src/rewrite/expr/convert.rs:569]: Why the change here from `Iterable` to `Array`? I'm assuming `Iterable` is a superset of the other, so what gets left out?

File: c2rust-analyze/src/rewrite/expr/convert.rs [hunk 4]
```diff
@@ -663,6 +797,33 @@ pub fn convert_cast_rewrite(kind: &mir_op::RewriteKind, hir_rw: Rewrite) -> Rewr
...
+            Rewrite::Call(
+                "std::mem::replace".to_string(),
+                vec![
+                    Rewrite::Ref(Box::new(hir_rw), hir::Mutability::Mut),
+                    Rewrite::Text("Err(())".into()),
+                ],
+            )
+        }
+        mir_op::RewriteKind::DynOwnedWrap => {
+            Rewrite::Call("std::result::Result::<_, ()>::Ok".to_string(), vec![hir_rw])
```
Comment (ahomescu) [inline c2rust-analyze/src/rewrite/expr/convert.rs:814]: Couldn't this just be an `Option`?

File: c2rust-analyze/src/rewrite/expr/convert.rs [hunk 5]
```diff
@@ -286,6 +286,119 @@ impl<'tcx> ConvertVisitor<'tcx> {
...
+                        ));
+                        stmts.push(format_rewrite!(
+                            "dest_ptr.resize_with(dest_n, || {})",
+                            zeroize_expr,
+                        ));
+                        Rewrite::Text("dest_ptr.into_boxed_slice()".into())
+                    }
+                    (false, true) => {
+                        format_rewrite!(
+                            "src_ptr.into_iter().next().unwrap_or_else(|| {})",
```
Comment (ahomescu) [inline c2rust-analyze/src/rewrite/expr/convert.rs:379]: Couldn't we use `unwrap_or_default()`?

File: c2rust-analyze/src/rewrite/expr/convert.rs (hunk 3 above)
Comment (ahomescu) [reply c2rust-analyze/src/rewrite/expr/convert.rs:569]: Answered my own question: it's just a rename.

File: c2rust-analyze/src/rewrite/expr/convert.rs [hunk 6]
```diff
@@ -594,6 +707,27 @@ fn generate_zeroize_code(zero_ty: &ZeroizeType, lv: &str) -> String {
     }
 }
 
+/// Generate an expression to produce a zeroized version of a value.
+fn generate_zeroize_expr(zero_ty: &ZeroizeType) -> String {
```
Comment (ahomescu) [inline c2rust-analyze/src/rewrite/expr/convert.rs:711]: Couldn't you use `Default::default` here? Is it because the trait isn't implemented for generic-sized arrays?

File: c2rust-analyze/src/rewrite/expr/convert.rs (hunk 2 above)
Comment (spernsteiner) [reply c2rust-analyze/src/rewrite/expr/convert.rs:326]: `vec![x; n]` requires `x` to be `Clone`, so it can't contain e.g. `Option<&mut T>`

File: c2rust-analyze/src/rewrite/expr/convert.rs (hunk 3 above)
Comment (spernsteiner) [reply c2rust-analyze/src/rewrite/expr/convert.rs:569]: It was previously called `Iterable` because it could in principle be used for anything with an `iter_mut()` method.  It's now called `Array` because it can only be used for arrays specifically - the `generate_zeroize_expr` function used for initializing allocations can't be generic as the code we use for zeroizing an existing value on `memset(_, 0, _)`.

File: c2rust-analyze/src/rewrite/expr/convert.rs (hunk 4 above)
Comment (spernsteiner) [reply c2rust-analyze/src/rewrite/expr/convert.rs:814]: It could, but `DynOwned` can be composed with `Option<T>` if we encounter a nullable owned pointer, and `Option<DynOwned<T>>` is clearer than `Option<Option<T>>`.  But c2rust-analyze doesn't yet have a run-time support library where we could define `DynOwned<T>`, so for now we use the somewhat ugly `Result<T, ()>` in place of `DynOwned<T>`.

File: c2rust-analyze/src/rewrite/expr/convert.rs (hunk 6 above)
Comment (spernsteiner) [reply c2rust-analyze/src/rewrite/expr/convert.rs:711]: It's also not implemented for c2rust-transpiled structs

File: c2rust-analyze/src/rewrite/expr/convert.rs (hunk 1 above)
Comment (spernsteiner) [reply c2rust-analyze/src/rewrite/expr/convert.rs:299]: No, but there's a related issue of potentially using the wrong size when converting the byte count passed to `malloc`/`memcpy`/`memset` to an element count.  I have a fix for this (at least for the most common cases) in a later branch: fd24d8158dbd7e4f0900661ae8a6afe4bc4bbc38

**Comment by spernsteiner** [comment]
Looks like the Darwin CI error here is the same issue that's being fixed in #1146.  Once that's merged, I'll rebase this and try the CI again.

**Comment by thedataking** [comment]
> Looks like the Darwin CI error here is the same issue that's being fixed in #1146. Once that's merged, I'll rebase this and try the CI again.

You are right. #1146 is merged now.

File: c2rust-analyze/src/rewrite/expr/convert.rs (hunk 4 above)
Comment (ahomescu) [reply c2rust-analyze/src/rewrite/expr/convert.rs:812]: Might be useful to add a comment.

**Review by ahomescu — APPROVED** [review]

File: c2rust-analyze/src/rewrite/expr/convert.rs (hunk 4 above)
Comment (spernsteiner) [reply c2rust-analyze/src/rewrite/expr/convert.rs:812]: Added a comment on the translation of `DynOwned` types: 95d40382e5f9f71f4010f7ed211239fc585d6343
